app/scripts/migrations/151.ts

```diff
@@ -0,0 +1,308 @@
+        'Skipping migration: No chain ID was found, or the chain ID resolved from the `NetworkController.selectedNetworkClientId` value is not a string.',
+      ),
+    );
+
+    return state;
+  }
+
+  const updatedSubjects: string[] = [];
+
+  // Add permission to use the current globally selected network to all Snaps
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration = permissionKeys.includes(
+      SnapEndowments.EthereumProvider,
+    );
+
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caveatValue = addPermittedEthChainId(
+      getCaveat(subject),
+      currentChainId,
+    );
+
+    const {
+      date = Date.now(),
+      id = nanoid(),
+      invoker = key,
+      parentCapability = Caip25EndowmentPermissionName,
+    } = subject.permissions[Caip25EndowmentPermissionName] ?? {};
+
+    const newSubject: GenericPermissionControllerSubject = {
+      ...subject,
+      permissions: {
+        ...subject.permissions,
+        [Caip25EndowmentPermissionName]: {
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: caveatValue,
+            },
+          ],
+          date,
+          id,
+          invoker,
+          parentCapability,
+        },
+      },
+    };
+
+    return {
+      ...accumulator,
+      [key]: newSubject,
+    };
+  }, {});
+
+  const currentDomains = selectedNetworkControllerState.domains;
+  const domains = Object.fromEntries(
+    updatedSubjects.map((subject) => [subject, currentChainId]),
+  );
+
+  selectedNetworkControllerState.domains = {
+    ...currentDomains,
+    ...domains,
+  };
+
+  return state;
+}
+
+/**
+ * Get the existing `endowment:caip25` caveat value for a given subject, if any.
+ * If the caveat does not exist, a new empty caveat value is returned.
+ *
+ * This function does not validate the structure of the existing caveat value,
+ * and assumes that the existing value is valid.
+ *
+ * @param subject - The subject to get the existing caveat value for.
+ * @returns The existing `endowment:caip25` caveat value.
+ */
+function getCaveat(
+  subject: GenericPermissionControllerSubject,
+): Caip25CaveatValue {
+  const existingCaveat = subject.permissions[
+    Caip25EndowmentPermissionName
+  ]?.caveats?.find((caveat) => caveat.type === Caip25CaveatType);
+
+  if (!existingCaveat) {
+    return {
+      isMultichainOrigin: false,
+      optionalScopes: {},
+      requiredScopes: {},
+    };
+  }
+
+  return existingCaveat.value as Caip25CaveatValue;
```

Comment:
Just curious how could this ever be hit?

Reply:
And if it is hit should we check what chainId/scopes are already set in the caveat?

Reply:
From my understanding `eth_accounts` uses the same caveat, does it not?

Reply:
yes it does now...  I'm forgetting though have we already granted snaps the caip25 permission elsewhere?

Reply:
Not to my knowledge, I think we only use it for accounts.

Reply:
I thought this was the first instance of us giving snaps this new permission?

Reply:
sorry nvm